lib/spack/spack/environment.py

```diff
@@ -557,7 +557,8 @@ def regenerate(self, all_specs, roots):
             if not os.path.isabs(root):
                 root = os.path.normpath(os.path.join(self.base, self.root))
             fs.mkdirp(root)
-            with fs.replace_directory_transaction(root):
+
+            with fs.replace_directory_transaction(root, os.path.dirname(root)):
```

Comment:
Firstly, this needs to add a comment explaining that the default tmpdir chosen for storing the temporary files may not work.

Also, if the tmpdir is not able to create valid symlinks to the install, then how are Spack installs succeeding? Aren't the stages typically in a tmpdir? I assume to fix this the user would update the config to move the tmpdir, but in that case, we would ideally use that same working tmpdir instance here (it might have qualities desired of a tmpdir).

Reply:
```suggestion
            # The tempdir for the directory transaction must be in the same 
            # filesystem mount as the view, so that view creation fails to 
            # symlink only when the finished view would fail as well
            with fs.replace_directory_transaction(root, os.path.dirname(root)):
```